### PR TITLE
executor: fix wrong index hash join when hash col is enum

### LINKED ISSUE #27893: inl_merge_join and inl_hash_join have different results
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

use test;
source [test.txt](https://github.com/pingcap/tidb/files/7128387/test.txt)

```
select /*+ inl_merge_join(t1) */ count(*) from table_400_undef_undef t1 right join table_400_undef_undef t2 on t1. `col_int_key_signed` = t2. `col_enum_key_signed` and t1. `col_char(20)_key_signed` < t2. `col_char(20)_key_signed` where t1. `col_bigint_key_signed` != 53 order by t1.pk, t2.pk;

select /*+ inl_hash_join(t1) */ count(*) from table_400_undef_undef t1 right join table_400_undef_undef t2 on t1. `col_int_key_signed` = t2. `col_enum_key_signed` and t1. `col_char(20)_key_signed` < t2. `col_char(20)_key_signed` where t1. `col_bigint_key_signed` != 53 order by t1.pk, t2.pk;
```
### 2. What did you expect to see? (Required)
```
mysql> select /*+ inl_merge_join(t1) */ count(*) from table_400_undef_undef t1 right join table_400_undef_undef t2 on t1. `col_int_key_signed` = t2. `col_enum_key_signed` and t1. `col_char(20)_key_signed` < t2. `col_char(20)_key_signed` where t1. `col_bigint_key_signed` != 53 order by t1.pk, t2.pk;
+----------+
| count(*) |
+----------+
|       97 |
+----------+
1 row in set, 1 warning (0.00 sec)

mysql>
mysql> select /*+ inl_hash_join(t1) */ count(*) from table_400_undef_undef t1 right join table_400_undef_undef t2 on t1. `col_int_key_signed` = t2. `col_enum_key_signed` and t1. `col_char(20)_key_signed` < t2. `col_char(20)_key_signed` where t1. `col_bigint_key_signed` != 53 order by t1.pk, t2.pk;
+----------+
| count(*) |
+----------+
|       97 |
+----------+
1 row in set, 1 warning (0.00 sec)
```
### 3. What did you see instead (Required)
```
mysql> select /*+ inl_merge_join(t1) */ count(*) from table_400_undef_undef t1 right join table_400_undef_undef t2 on t1. `col_int_key_signed` = t2. `col_enum_key_signed` and t1. `col_char(20)_key_signed` < t2. `col_char(20)_key_signed` where t1. `col_bigint_key_signed` != 53 order by t1.pk, t2.pk;
+----------+
| count(*) |
+----------+
|       97 |
+----------+
1 row in set, 2 warnings (0.13 sec)

mysql>
mysql> select /*+ inl_hash_join(t1) */ count(*) from table_400_undef_undef t1 right join table_400_undef_undef t2 on t1. `col_int_key_signed` = t2. `col_enum_key_signed` and t1. `col_char(20)_key_signed` < t2. `col_char(20)_key_signed` where t1. `col_bigint_key_signed` != 53 order by t1.pk, t2.pk;
+----------+
| count(*) |
+----------+
|        0 |
+----------+
1 row in set (0.05 sec)
```
### 4. What is your TiDB version? (Required)

```
Release Version: v5.2.0-3-gcd8fb24c5
Edition: Community
Git Commit Hash: cd8fb24c5f7ebd9d479ed228bb41848bd5e97445
Git Branch: release-5.2
UTC Build Time: 2021-09-08 10:01:42
GoVersion: go1.16.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by lzmhhh123**
Signed-off-by: lzmhhh123 <lzmhhh123@gmail.com>

### What problem does this PR solve?

Issue Number: close #27893

### What is changed and how it works?

What's Changed:

How it Works:

### Check List 

Tests 

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Fix wrong index hash join when hash col is enum.
```

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -2608,6 +2608,7 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plannercore.PhysicalIndexJoin)
 	for i, col := range v.OuterHashKeys {
 		outerTypes[col.Index] = outerTypes[col.Index].Clone()
 		outerTypes[col.Index].Collate = innerTypes[v.InnerHashKeys[i].Index].Collate
+		outerTypes[col.Index].Flag = col.RetType.Flag
```
Comment (wjhuang2016) [inline executor/builder.go:2611]: Why do we need this?

Comment (lzmhhh123) [reply executor/builder.go:2611]: `EnumSetAsIntFlag`

Comment (wjhuang2016) [reply executor/builder.go:2611]: Ok, please update the comment.

**Review by qw4990 — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**